File: iot/amqp/AMQPParser.py
import iot.amqp.avps.HeaderCode as HeaderCode
import iot.amqp.header.api.HeaderFactory as HeaderFactory
import iot.amqp.header.impl.AMQPPing as AMQPPing
import iot.amqp.header.impl.AMQPProtoHeader as AMQPProtoHeader
import iot.amqp.numeric.NumericUtil as NumericUtil

class amqpParser():
    def __init__(self):
        self.index = 0
    
    def next(self, data, index):
        length = NumericUtil.util.getInt(data[self.index:self.index + 4])
        self.index = index + 4
        if length == 1095586128:
            protocolId = NumericUtil.util.getByte(data, self.index)
            self.index += 1
            versionMajor = NumericUtil.util.getByte(data, self.index)
            self.index += 1
            versionMinor = NumericUtil.util.getByte(data, self.index)
            self.index += 1
            versionRevision = NumericUtil.util.getByte(data, self.index)
            self.index += 1
            if (protocolId == 0 or protocolId == 3) and versionMajor == 1 and versionMinor == 0 and versionRevision == 0:
                self.index = 0
                return data[index: 8]

        self.index = 0
        return data[index:length]
    
    def encode(self, header):
        buf = None

        if header.toStringShort() == 'AMQPProtoHeader':
            buf = bytearray()
            buf = NumericUtil.util.addString(buf,'AMQP')
            buf = NumericUtil.util.addByte(buf,header.getProtocolId())
            buf = NumericUtil.util.addByte(buf, header.getVersionMajor())
            buf = NumericUtil.util.addByte(buf, header.getVersionMinor())
            buf = NumericUtil.util.addByte(buf, header.getVersionRevision())
            return buf

        if header.toString() == 'AMQPPing':
            buf = bytearray()
            buf = NumericUtil.util.addInt(buf,8)
            buf = NumericUtil.util.addByte(buf,header.getDoff())
            buf = NumericUtil.util.addByte(buf, header.getType())
            buf = NumericUtil.util.addShort(buf, header.getChannel())
            return buf

        length = 8
        arguments = header.toArgumentsList()
        length += arguments.getLength()

        sections = None
        headerCode = HeaderCode.headerCode()
        if header.getCode() == headerCode.getValueByKey('TRANSFER'):
            sections = header.getSections()
            if sections is not None:
                for section in sections.values():
                    length += section.getValue().getLength()

        buf = bytearray()
        buf = NumericUtil.util.addInt(buf,length)
        buf = NumericUtil.util.addByte(buf,header.getDoff())
        buf = NumericUtil.util.addByte(buf, header.getType())
        buf = NumericUtil.util.addShort(buf, header.getChannel())

        if arguments.toString() == 'TLVList':
            buf += arguments.getBytes()

        if sections is not None:
            for section in sections.values():
                value = section.getValue()
                buf += value.getBytes()

        self.index = 0
        return buf
    
    def decode(self, data):
        length = NumericUtil.util.getInt(data[self.index:self.index+4])
        self.index += 4
        doff = NumericUtil.util.getByte(data,self.index) & 0xff
        self.index += 1
        type = NumericUtil.util.getByte(data,self.index) & 0xff
        self.index += 1
        channel = NumericUtil.util.getShort(data[self.index:self.index+2]) & 0xffff
        self.index += 2

        if length == 8 and doff == 2 and (type == 0 or type == 1) and channel == 0:
            if self.index >= len(data):
                self.index = 0
                return AMQPPing.amqpPing()
            else:
                print("Received malformed ping-header with invalid length")

        if length == 1095586128 and (doff == 3 or doff == 0) and type == 1 and channel == 0:
            if self.index >= len(data):
                self.index = 0
                return AMQPProtoHeader.amqpProtoHeader(doff)
            else:
                print("Received malformed protocol-header with invalid length")
        header = None
        
        headerFactory = HeaderFactory.headerFactory(self.index)
        if type == 0:
            header = headerFactory.getAMQP(data)
        elif type == 1:
            header = headerFactory.getSASL(data)
        else:
            print("Received malformed header with invalid type: " + type)

        self.index = headerFactory.getIndex()
        
        header.setDoff(doff)
        header.setType(type)
        header.setChannel(channel)
        headerCode = HeaderCode.headerCode()
        if header.getCode() == headerCode.getValueByKey('TRANSFER'):
            header.setSections({})
            while self.index < len(data):
                headerFactory.setIndex(self.index)
                section = headerFactory.getSection(data)
                header.sections[section.getCode()] = section
                self.index = headerFactory.getIndex()
        
        self.index = 0
        return header

File: IoTbrokerClient/src/iot-protocols/amqp/classes/parser/amqpparser.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpparser.h"

AMQPParser::AMQPParser()
{

}

int AMQPParser::next(unsigned char * buffer)
{
    int i = 0;

    int length = get_int(buffer, i);
    i += 4;

    if (length == 1095586128) {
        int protocolId = buffer[i++];
        int versionMajor = buffer[i++];
        int versionMinor = buffer[i++];
        int versionRevision = buffer[i];
        if ((protocolId == AMQPProtocolId || protocolId == AMQPProtocolIdTLS || protocolId == AMQPProtocolIdSASL) &&
                versionMajor == 1 && versionMinor == 0 && versionRevision == 0) {
            return 8;
        } else {
            return -1;
        }
    }
    return length;
}

AMQPHeader *AMQPParser::decode(ByteArray buffer)
{
    int i = 0;
    String protocolName;
    protocolName = get_string(&buffer[i], 4);
    i += 4;
    if (protocolName == AMQPProtocolName) {
        AMQPProtoHeader *protoHeader = new AMQPProtoHeader();
        protoHeader->setProtocolId((AMQPProtocolIds)buffer[i++]);
        protoHeader->setVersionMajor(buffer[i++]);
        protoHeader->setVersionMinor(buffer[i++]);
        protoHeader->setVersionRevision(buffer[i++]);
        return protoHeader;
    }

    int length = get_int(buffer, i) & 0xffffffffL;
    i += 4;
    int doff = buffer[i++] & 0xff;
    int type = buffer[i++] & 0xff;
    int channel = get_short(buffer, i) & 0xffff;
    i += 2;

    if (length == 8 && doff == 2 && (type == 0 || type == 1) && channel == 0) {
        if (sizeof(buffer) - i == 0) {
            return new AMQPPing();
        } else {
            printf("Received malformed ping-header with invalid length");
            return NULL;
        }
    }

    if (length == 1095586128 && (doff == 3 || doff == 0) && type == 1 && channel == 0) {
        if (sizeof(buffer) - i == 0) {
            return new AMQPPing();
        } else {
            printf("Received malformed protocol-header with invalid length");
            return NULL;
        }
    }

    if (length != (sizeof(buffer) - i) + 8) {
        printf("Received malformed header with invalid length");
        return NULL;
    }

    AMQPHeader *header = NULL;

    if (type == 0) {
        header = AMQPFactory::amqp(&buffer);
    } else if (type == 1) {
        header = AMQPFactory::sasl(&buffer);
    } else {
        printf("Received malformed header with invalid type");
        return NULL;
    }

    if (header->getCode()->getValue() == AMQP_TRANSFER_HEADER_CODE) {
        while ((sizeof(buffer) - i) > 0) {
            list<AMQPSection *> sections;
            sections.push_back(AMQPFactory::section(&buffer));
            ((AMQPTransfer *)header)->addSections(sections);
        }
    }

    return header;
}

unsigned char * AMQPParser::encode(AMQPHeader *header)
{
    unsigned char * data = NULL;
    int i = 0;
    int string_length = 0;

    if (header->getCode()->getValue() == AMQP_PROTOCOL_HEADER_CODE_STRING) {
        string_length = strlen(AMQPProtocolName);
        buf = malloc(string_length * sizeof(char));
        memcpy(&buf[i], AMQPProtocolName, string_length);
        i += string_length;
        data[i++] = ((AMQPProtoHeader *)header)->getProtocolId();
        data[i++] = ((AMQPProtoHeader *)header)->getVersionMajor();
        data[i++] = ((AMQPProtoHeader *)header)->getVersionMinor();
        data[i++] = ((AMQPProtoHeader *)header)->getVersionRevision();
        return data;
    }

    if (header->getCode()->getValue() == AMQP_PING_HEADER_CODE_STRING) {
        int ping_length = 8;
        buf = malloc(ping_length * sizeof(char));
        i += add_int(&data[i], 8);
        data[i++] = header->getDoff();
        data[i++] = header->getType();
        add_short(&buf[i], header->getChannel());
        return data;
    }

    int frame_length = 8;
    data = malloc((frame_length + header->arguments()->getLength()) * sizeof(char));

    i += add_int(&data[i], header->getLength());
    data[i++] = header->getDoff();
    data[i++] = header->getTypeValue();
    i += add_short(&buf[i], header->getChannel());

    string_length = strlen(header->arguments()->getData());
    memcpy(&data[i], header->arguments()->getData(), string_length);
    i += string_length;

    if (header->getCode()->getValue() == AMQP_TRANSFER_HEADER_CODE) {
      vector<AMQPSection *> sections;
      map<AMQPSectionCode *, AMQPSection *> sectionsMap = (AMQPTransfer *)header)->getSections();
      for(map<AMQPSectionCode  *, AMQPSection *>::iterator it = sections.begin(); it != sections.end(); ++it) {
          sections.push_back(it->second);
      }
      foreach (AMQPSection *item, sections) {
          string_length = sizeof(item->getValue()->getData());
          memcpy(&data[i], item->getValue()->getData(), string_length);
          i += string_length;
      }
    }

    return data;
}
